google-cloud-firestore/lib/google/cloud/firestore/aggregate_query.rb

```diff
@@ -107,12 +111,25 @@ def initialize query, parent_path, client, aggregates: []
         #
         def add_count aggregate_alias: nil
           aggregate_alias ||= ALIASES[:count]
-          new_aggregates = @aggregates.dup
-          new_aggregates << StructuredAggregationQuery::Aggregation.new(
-            count: StructuredAggregationQuery::Aggregation::Count.new,
+          new_aggregate = Google::Cloud::Firestore::V1::StructuredAggregationQuery::Aggregation.new(
+            count: Google::Cloud::Firestore::V1::StructuredAggregationQuery::Aggregation::Count.new,
             alias: aggregate_alias
           )
-          AggregateQuery.start query, new_aggregates, parent_path, client
+
+          start new_aggregate
+        end
+
+        ##
+        # @private
+        def start new_aggregate
+          combined_aggregates = [].concat(grpc.aggregations).concat([new_aggregate])
+          new_grpc = Google::Cloud::Firestore::V1::StructuredAggregationQuery.new(
+            structured_query: @query,
+            aggregations: combined_aggregates
+          )
+          dup.tap do |aq|
+            aq.instance_variable_set :@grpc, new_grpc
```

Comment:
I'm actually kind of puzzled by this refactor. It seems to make things a bit more awkward, since we now have to mutate the object (and use `instance_variable_set`). What are we trying to accomplish here?

Reply:
Basically, the classes in Firestore are (mostly) designed to prefer immutable objects. When user adds something to an object (in this case, an _aggregate_), we first clone the existing the object, and change the duplicate object.

This pattern of `instance_variable_set` seems to be followed across many gems in this repo ([example](https://github.com/googleapis/google-cloud-ruby/blob/0689c884e80e483248c4e489a32d3bb93c5009e3/google-cloud-firestore/lib/google/cloud/firestore/transaction.rb#L693), [example](https://github.com/googleapis/google-cloud-ruby/blob/0689c884e80e483248c4e489a32d3bb93c5009e3/google-cloud-bigtable/lib/google/cloud/bigtable/column_family_map.rb#L347) and [example](https://github.com/googleapis/google-cloud-ruby/blob/0689c884e80e483248c4e489a32d3bb93c5009e3/google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb#L1434)). So I'm following the same pattern.

Reply:
Yes, I agree, it's preferred to use immutable objects whenever possible. I think the previous code already did that. This change actually seems to back away from immutability a bit, by using `dup` (which can be risky because it bypasses the initialize method) and by mutating the instance variable after construction. Yes, there are other cases of using `instance_variable_set` elsewhere in the repo for constructing objects. It's an antipattern and really should be done differently. Normally the best way to construct immutable objects is the simplest: by passing the desired state into the constructor just like the old version of this code did.

Reply:
There's actually a problem with my previous code as well 😅. The constructor was allowing the `aggregates` param, which really isn't the job of any external class to pass it. I wanted to fix this anti-pattern, & hide the field from other classes.

### `dup` method:

Thanks for the feedback on `dup`. I modified the code to rely on the constructor, instead of `dup`.

### `instance_variable_set` method: 

I still kept this usage of `instance_variable_set`, because the only other way around it is to pass the `aggregates` (or `@grpc`) as a param in the constructor (whose issue I listed at the start). I can change this if you feel it should be the other way around. But, compromising between "sacrificing abstraction" and "sacrificing true immutable code", I'm leaning slightly towards the latter option.